docs/design-proposals/multi-tenancy/multi-tenancy-access-control.md

```diff
@@ -0,0 +1,181 @@
+
+## Authentication
+
+Same to Kubernetes, Arktos supports configurable multiple authenticators. Each authenticator implements a different way of authenticating users or service accounts. For an incoming API request, each authenticator will be tried in sequence until one of them succeed. 
+
+The key changes to each authenticator is that it needs to retrieve the tenant information and put it into request context.
+
+### 1.Certificate Authenticator
+
+The existing X509 certificate authenticator reads "/CN" and "/O" from cert subject, and map them to user name and groups respectively. 
+
+To make the authenticator tenancy-aware, "/O" will be mapped to tenant while "/OU" (Organization Unit) will be mapped to groups. 
+
+Below is a sample cert:
+
+```text
+openssl req -new -key demo.pem -out demo-csr.pem -subj "/CN=demo/O=tenant1/OU=app1/OU=app2"
+```
+
+The certificate authenticator will map it to the following user:
+* The user name is "demo".
+* The user belongs to tenant "tenant1".
+* The user has a membership of group "app1" and "app1".
+
+### 2.Token Authenticator
+(**TBD**)
+
+### 3.Basic Authenticator
+(**TBD**)
+
+### 4.Service Account Authenticator
+(**TBD**)
+
+## Authorization
+
+Arktos supports different ways of authorization: RBAC, ABAC, Node and Webhook.
+
+In general the authorization rule is:
+
+* A system tenant user can access any space.
+* A non-system tenant user can only access its own space.
+* Optionally, a non-system tenant admin can create a policy object in its space to allow other tenants to access its space. The policy object will specify which tenants are allowed to access which resources.
+
+### 1.RBAC Authorizer
+
+The following new check are added into the RBAC implementation:
+
+```go
+if (user.tenant != "system") && (user.tenant != resource.tenant) {
+   return DecisionDeny
+}
+```
+
+Any cross-tenant attempt not issued by a system tenant user will be blocked here. If this check passes, it continues to the original RBAC logics:
+
+```go
+If any ClusterRoleBinding allows the request {
+   return DecisionAllow
+}
+
+If the requested resource is a namespace-scoped resource {
+   if any RoleBinding within the namespace allows the request {
+      return DecisionAllow
+   }
+}
+
+return DecisionDeny
+```
+
+With this change, all existing RBAC policies still work.  
+
+### 2.ABAC Authorizer
+
+The ABAC authorizer needs to be changed to read in an extra field, "tenant", from the ABAC policy file. And also compare the user.tenant with requestInfo.tenant in its authorization check.
+
+Below is a sample ABAC policy file that says user Alice in tenant A can access resources tenants/tenant-a/* :
+
+```go
+{"apiVersion": "abac.authorization.kubernetes.io/v1beta1", "kind": "Policy", "spec": {"tenant": "tenant A", "user": "alice", "tenant": "tenant-a", "namespace": "*", "resource": "*", "apiGroup": "*"}}
+```
+
+If the tenant field is missing, it is the system tenant. If the space field is missing, it's a resource in system space.
```

Comment:
if the tenant field is missing by some non-system tenant by mistake or on purpose, should we treat this as a system tenant?

Reply:
This file is configured by cluster admin and the file location is provided as a start-up parameter to the API server.